app/controllers/sentences_controller.php

```diff
@@ -216,8 +217,21 @@ public function show($id = null)
 
             $randomId = $this->Sentence->getRandomId($lang);
 
-            $this->Session->write('random_lang_selected', $lang);
-            $this->redirect(array("action"=>"show", $randomId));
+            if (is_null($randomId)) {
+                $searchDisabled = !Configure::read('Search.enabled');
+                $this->set('searchProblem', true);
+                if ($searchDisabled) {
+                    $this->Session->setFlash(__('Search was disabled. Please try again later.', true));
+                } else {
+                    $html = new HtmlHelper();
+                    $this->Session->setFlash(format(__('An error occurred while performing this function. 
+                            If the error persists, please <a href="{}">let us know</a>.', true),
+                        $html->url(array("controller"=>"pages", "action" => "contact"))));
```

Comment:
You’re messing with MVC hard here. :-)

One way to avoid this is to print the message directly from the view, instead of using flash messages. See this [example](https://github.com/Tatoeba/tatoeba2/blob/dev/app/views/transcriptions/view.ctp).